pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,120 @@
+from datetime import timedelta
+import numbers
+from typing import (
+    Any,
+    Generic,
+    TypeVar,
+    overload,
+)
+
+import numpy as np
+import numpy.typing as npt
+
+from pandas._libs import (
+    Timedelta,
+    Timestamp,
+)
+from pandas._typing import IntervalBound
+
+_OrderableMixinT = TypeVar(
+    "_OrderableMixinT", int, float, Timestamp, Timedelta, npt.NDArray[np.generic]
+)
+_OrderableT = TypeVar("_OrderableT", int, float, Timestamp, Timedelta)
+
+# note: mypy doesn't support overloading properties
+# based on github.com/microsoft/python-type-stubs/pull/167
+class _LengthProperty:
+    @overload
+    def __get__(self, instance: IntervalMixin[Timestamp], owner: Any) -> Timedelta: ...
+    @overload
+    def __get__(
+        self, instance: IntervalMixin[_OrderableMixinT], owner: Any
+    ) -> _OrderableMixinT: ...
+
+class IntervalMixin(Generic[_OrderableMixinT]):
+    @property
+    def closed_left(self) -> bool: ...
+    @property
+    def closed_right(self) -> bool: ...
+    @property
+    def open_left(self) -> bool: ...
+    @property
+    def open_right(self) -> bool: ...
+    @property
+    def mid(self) -> _OrderableT: ...
+    length: _LengthProperty
+    @property
+    def is_empty(self) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...) -> None: ...
+
+class Interval(IntervalMixin[_OrderableT]):
+    def __init__(
+        self,
+        left: _OrderableT,
+        right: _OrderableT,
+        closed: IntervalBound = ...,
+    ) -> None: ...
+    @property
+    def closed(self) -> str: ...
+    @property
+    def left(self) -> _OrderableT: ...
+    @property
+    def right(self) -> _OrderableT: ...
+    def __str__(self) -> str: ...
+    # TODO: could return Interval with different type
+    def __add__(
+        self, y: numbers.Number | np.timedelta64 | timedelta
```

Comment:
I think you need to make the operators type specific.  For `Interval[Timestamp]`, you can only add `Timedelta`.  For the numeric ones, you can only add `float` or `int`.  See the latest in https://github.com/microsoft/python-type-stubs/pull/167/